qiskit/circuit/commutation_checker.py

```diff
@@ -201,7 +214,60 @@ def _hashable_parameters(params):
     return ("fallback", str(params))
 
 
-_skipped_op_names = {"measure", "reset", "delay"}
+def is_commutation_supported(op):
+    """
+    Filter operations whose commutation is not supported due to bugs in transpiler passes invoking
+    commutation analysis.
+    Args:
+        op (Operation): operation to be checked for commutation relation
+    Return:
+        True if determining the commutation of op is currently supported
+    """
+    # Bug in CommutativeCancellation, e.g. see gh-8553
+    if getattr(op, "condition", False):
+        return False
+
+    # Commutation of ControlFlow gates also not supported yet. This may be pending a control flow graph.
+    if op.name in CONTROL_FLOW_OP_NAMES:
+        return False
+
+    return True
+
+
+def is_commutation_skipped(op, qargs, max_num_qubits):
+    """
+    Filter operations whose commutation will not be determined.
+    Args:
+        op (Operation): operation to be checked for commutation relation
+        qargs (List): operation qubits
+        max_num_qubits (int): the maximum number of qubits to consider, the check may be skipped if
+                the number of qubits for either operation exceeds this amount.
+    Return:
+        True if determining the commutation of op is currently not supported
+    """
+    if (
+        len(qargs) > max_num_qubits
+        or getattr(op, "_directive", False)
+        or op.name in _skipped_op_names
+    ):
+        return True
+
+    if getattr(op, "is_parameterized", False) and op.is_parameterized():
+        return True
+
+    # we can proceed if op has defined: to_operator, to_matrix and __array__, or if its definition can be
+    # recursively resolved by operations that have a matrix. We check this by constructing an Operator.
+    if (hasattr(op, "to_matrix") and hasattr(op, "__array__")) or hasattr(op, "to_operator"):
+        return False
+
+    # last resort: try to generate an Operator out of op, if this succeeds we can determine commutativity
+    # this might be runtime-intensive in general so we should attempt to branch out before running this
+    # it may be more runtime-efficient to check for entries in the commutation library before
+    try:
+        Operator(op)
```

Comment:
Yes, I agree. However, I would need to refactor the commutation checker quite a bit and I wanted to do this PR quickly. :-) Is it okay to leave it as is? I also suspect that this part of the function should not be executed often in practice.

Reply:
I think it's fine to leave it as is, but maybe put a TODO comment or something in there to indicate longer term it'd be better to de-duplicate the `Operator()` call.

Reply:
I shifted the `Operator()` resolution to `_commute_matmul` - while this would lead to unnecessary checks for operations that do not have a suitable `Operator` resolution, I would expect these checks to be comparatively lightweight.